File: lib/java/src/main/java/uapi/internal/generation/ConstructRandomUnion.java
package uapi.internal.generation;

import static uapi.internal.generation.ConstructRandomStruct.constructRandomStruct;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import uapi.RandomGenerator;
import uapi.internal.types.UStruct;

public class ConstructRandomUnion {
    static Map<String, Object> constructRandomUnion(Map<String, UStruct> unionCasesReference,
            Map<String, Object> startingUnion,
            boolean includeOptionalFields, boolean randomizeOptionalFields,
            RandomGenerator randomGenerator) {
        if (!startingUnion.isEmpty()) {
            final var entry = startingUnion.entrySet().stream().findAny().get();
            final var unionCase = entry.getKey();
            final var unionStructType = unionCasesReference.get(unionCase);
            final var unionStartingStruct = (Map<String, Object>) startingUnion.get(unionCase);

            return Map.of(unionCase, constructRandomStruct(unionStructType.fields, unionStartingStruct,
                    includeOptionalFields, randomizeOptionalFields, randomGenerator));
        } else {
            final var sortedUnionCasesReference = new ArrayList<>(unionCasesReference.entrySet());

            Collections.sort(sortedUnionCasesReference, (e1, e2) -> e1.getKey().compareTo(e2.getKey()));

            final var randomIndex = randomGenerator.nextIntWithCeiling(sortedUnionCasesReference.size() - 1);
            final var unionEntry = sortedUnionCasesReference.get(randomIndex);
            final var unionCase = unionEntry.getKey();
            final var unionData = unionEntry.getValue();

            return Map.of(unionCase,
                    constructRandomStruct(unionData.fields, new HashMap<>(), includeOptionalFields,
                            randomizeOptionalFields,
                            randomGenerator));
        }
    }
}

File: lib/py/uapi/internal/generation/ConstructRandomUnion.py
from typing import TYPE_CHECKING, cast

if TYPE_CHECKING:
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.types.UStruct import UStruct
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration


def construct_random_union(union_cases_reference: dict[str, 'UStruct'],
                           starting_union: dict[str, object],
                           include_optional_fields: bool,
                           randomize_optional_fields: bool,
                           random_generator: 'RandomGenerator') -> dict[str, object]:
    from uapi.internal.generation.ConstructRandomStruct import construct_random_struct

    if not starting_union:
        sorted_union_cases_reference = sorted(
            union_cases_reference.items(), key=lambda x: x[0])
        random_index = random_generator.next_int_with_ceiling(
            len(sorted_union_cases_reference) - 1)
        union_case, union_data = sorted_union_cases_reference[random_index]
        return {union_case: construct_random_struct(union_data.fields, {}, include_optional_fields,
                                                    randomize_optional_fields, random_generator)}
    else:
        union_case, union_starting_struct = cast(
            tuple[str, dict[str, object]], next(iter(starting_union.items())))
        union_struct_type = union_cases_reference[union_case]
        return {union_case: construct_random_struct(union_struct_type.fields, union_starting_struct,
                                                    include_optional_fields, randomize_optional_fields,
                                                    random_generator)}
